Proposal: Introduce bundles (#871)
open-telemetry/weaver: docs/proposals/bundles.md

# Bundles

This document is a design proposal to add the ability to define bundles of attributes in semconv.

## Use cases

There are a few different use cases which bundles will help us to achieve.

### Defining signals

By introducing bundles we are able to bundle a collection of attributes together into a single reusable item. 
These items have the primary goal of reducing the need to copy paste definitions while also reducing the maintaince effort.

An example of where this would be useful is if a `server` bundle was defined similiar to below:

```
bundle:
  - name: server
    brief: Defines a server
    note:
    attributes
      - ref: sever.address
      - ref: sever.port
```

Once it is defined, it can then be used on other signals to extend the definition.

```
span:
  - name: messaging.produced
    brief: Defines a span for sending a message via a message brooker such as rabbitmq
    note:

>     attributes

Reviewer: ```suggestion
    attributes:
```

      - ref: message.system.name
    bundles:
      - ref: server
```

In terms of the resolved schema, these bundles are blended seamlessly in to the definition.

```
span:
  - name: messaging.produced

>     brief: Defines a span for sending a message via a message brooker such as rabbitmq

Reviewer: ```suggestion
    brief: Defines a span for sending a message via a message broker such as RabbitMQ
```

    note:
    attributes
      - id: message.system.name
        ....

>       - id: sever.address

Reviewer: ```suggestion
      - id: server.address
```

        ....
      - id: sever.port
        ....
```

By adopting this approach we are improving the mantainability of the definitions,

> by reusing the definition rather than copy and pasting strucutures.

Reviewer: ```suggestion
by reusing the definition rather than copying and pasting structures.
```

### Extending signals

This use case is similiar to the previous except it is focussed on offering optional contextual extensions.

> An example of this extension would be Cloud Events which are only applicable if the application is using Cloud Events, 

Reviewer: Link to docs for what Cloud Events is here?

> hence the usage is contextual based upon cloud events being used.

Reviewer: These are capitalized earlier so capitalize here for consistency.

```suggestion
hence the usage is contextual based upon Cloud Events being used.
```
